src/cursor/cur_layered.c

```diff
@@ -736,197 +718,183 @@ __clayered_position_constituent(WT_CURSOR_LAYERED *clayered, WT_CURSOR *c, bool
+    /*
+     * The cursor is positioned, but `iter_flag` is not set so we cannot rely on alternate cursor
+     * and need to position it.
+     */
+    if (!F_ISSET(clayered, iter_flag)) {
+        __clayered_position_constituent(clayered, c_alternate, forward);
+    }
+
+    /* If the alternate cursor's key is equal to the current one, we should move it as well. */
+    if (F_ISSET(c_alternate, WT_CURSTD_KEY_INT)) {
+        WT_RET(__clayered_cursor_compare(clayered, c_alternate, c_current, &cmp));
+        if (cmp == 0)
+            WT_RET_NOTFOUND_OK(__clayered_constituent_iter(c_alternate, forward));
+    }
 
-    /* The key was found and current cursor is not set - update the current cursor. */
-    if (ret == 0 && cmp == 0 && clayered->current_cursor == NULL)
-        clayered->current_cursor = constituent;
+    /* Move the current cursor. */
+    WT_RET_NOTFOUND_OK(__clayered_constituent_iter(c_current, forward));
 
+done:
+    if (!F_ISSET(clayered, iter_flag)) {
+        F_CLR(clayered, WT_CLAYERED_ITERATE_PREV | WT_CLAYERED_ITERATE_NEXT);
+        F_SET(clayered, iter_flag);
+    }
     return (0);
 }
 
 /*
- * __clayered_next --
- *     WT_CURSOR->next method for the layered cursor type.
+ * __clayered_iterate --
+ *     Common function for moving a layered cursor to the next or previous position.
  */
 static int
-__clayered_next(WT_CURSOR *cursor)
+__clayered_iterate(WT_CURSOR_LAYERED *clayered, bool forward, uint32_t iter_flag)
 {
-    WT_CURSOR *alternate_cursor, *c;
-    WT_CURSOR_LAYERED *clayered;
     WT_DECL_RET;
-    WT_SESSION_IMPL *session;
-    int cmp;
-    bool deleted;
 
-    clayered = (WT_CURSOR_LAYERED *)cursor;
+    bool deleted = false;
+    WT_SESSION_IMPL *session = CUR2S(clayered);
+    WT_CURSOR *cursor = &clayered->iface;
 
-    CURSOR_API_CALL(cursor, session, ret, next, clayered->dhandle);
     __cursor_novalue(cursor);
     WT_ERR(__clayered_enter(clayered, false, false, true));
 
-    /* If we aren't positioned for a forward scan, get started. */
-    if (clayered->current_cursor == NULL || !F_ISSET(clayered, WT_CLAYERED_ITERATE_NEXT)) {
-        /* The order matters: check the ingest table before the stable one. */
-        WT_ERR(__clayered_iterate_constituent(clayered, clayered->ingest_cursor, true));
-        WT_ERR(__clayered_iterate_constituent(clayered, clayered->stable_cursor, true));
-        F_SET(clayered, WT_CLAYERED_ITERATE_NEXT);
-        F_CLR(clayered, WT_CLAYERED_ITERATE_PREV);
-
-        /* We just positioned *at* the key, now move. */
-        if (clayered->current_cursor != NULL)
-            goto retry;
-    } else {
-retry:
-        /* If there are multiple cursors on that key, move them forward. */
-        if (clayered->current_cursor == clayered->stable_cursor)
-            alternate_cursor = clayered->ingest_cursor;
-        else
-            alternate_cursor = clayered->stable_cursor;
+    do {
+        WT_ERR(__clayered_iterate_constituents(clayered, iter_flag, deleted));
```

Comment:
Can you add some comments in this function laying out what we are doing, It is not easily understandable what is going on

Reply:
I added a comment and a FIXME as it was discussed